File: src/jianzhi27.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* mirrorTree(TreeNode* root) {
        //所谓镜像，就是交换非叶节点的两个子节点
        //如果是叶节点或者是空节点，就不交换
        //不需要返回值
        //时间击败：100%，空间击败：70.17%
        mirrorTreeCore(root);
        return root;
    }

    void mirrorTreeCore(TreeNode * root){
        if(root == nullptr || (root->left == nullptr && root->right == nullptr)){
            return;
        }
        else{
            //否则，交换他们的子节点，无论是否为空
            TreeNode * temp = root->left;
            root->left = root->right;
            root->right = temp;
            //继续向两个子树进发
            if(root->left){
                mirrorTreeCore(root->left);
            }
            if(root->right){
                mirrorTreeCore(root->right);
            }
            return;
        }
    }
};

File: src/jianzhi27.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def mirrorTree(self, root: TreeNode) -> TreeNode:
        #干翻蚂蚁链
        #递归，交换左右节点
        #两分钟，一次提交，时间击败-96.00%，空间击败-33.28%
        #递归边界
        if root == None:
            return None
        else:
            root.left, root.right = root.right, root.left
            self.mirrorTree(root.left)
            self.mirrorTree(root.right)
            return root
